File: tools/qc/cgm-ingestion-evidence/harnesses/scan.js
// Count live console.* call sites in lib/ + index.js, and how many pass a
// non-literal argument. Comments and string literals stripped first.
const fs = require('fs'), path = require('path');
function walk(d, out) {
  for (const e of fs.readdirSync(d, {withFileTypes:true})) {
    const p = path.join(d, e.name);
    if (e.isDirectory()) walk(p, out);
    else if (e.name.endsWith('.js')) out.push(p);
  }
  return out;
}
// strip // and /* */ comments, respecting string/template literals
function strip(src) {
  let out = '', i = 0, n = src.length;
  while (i < n) {
    const c = src[i], d = src[i+1];
    if (c === '/' && d === '/') { while (i < n && src[i] !== '\n') i++; continue; }
    if (c === '/' && d === '*') { i += 2; while (i < n && !(src[i] === '*' && src[i+1] === '/')) i++; i += 2; continue; }
    if (c === '"' || c === "'" || c === '`') {
      const q = c; out += c; i++;
      while (i < n) { if (src[i] === '\\') { out += 'XX'; i += 2; continue; }
        if (src[i] === q) break; out += (src[i] === '\n' ? '\n' : 'X'); i++; }
      out += q; i++; continue;
    }
    out += c; i++;
  }
  return out;
}
const root = process.argv[2];
const files = [];
if (fs.existsSync(path.join(root,'index.js'))) files.push(path.join(root,'index.js'));
if (fs.existsSync(path.join(root,'lib'))) walk(path.join(root,'lib'), files);
let live = 0, dyn = 0; const per = {};
for (const f of files) {
  const s = strip(fs.readFileSync(f,'utf8'));
  let l = 0, dd = 0;
  const re = /console\.(log|error|warn|info|debug)\s*\(/g; let m;
  while ((m = re.exec(s))) {
    l++;
    // read the balanced argument list
    let i = re.lastIndex, depth = 1, args = '';
    while (i < s.length && depth > 0) {
      if (s[i] === '(') depth++;
      else if (s[i] === ')') { depth--; if (!depth) break; }
      args += s[i]; i++;
    }
    // non-literal = anything that is not purely quoted strings / numbers joined by commas
    const cleaned = args.replace(/"[^"]*"|'[^']*'|`[^`]*`/g, 'S').replace(/\s+/g,'');
    if (!/^(S|[0-9.]+)(,(S|[0-9.]+))*,?$/.test(cleaned) && cleaned.length) dd++;
  }
  if (l) { per[path.relative(root,f)] = l + '/' + dd; live += l; dyn += dd; }
}
console.log(root.split('/').pop(), 'TOTAL', live + '/' + dyn);
for (const k of Object.keys(per).sort()) console.log('   ', k, per[k]);

File: tools/qc/cgm-ingestion-evidence/harnesses/scan.py
import os,re,sys,json

def strip(src):
    # returns source with comments blanked (keep newlines) and string literals replaced by ""
    out=[]; i=0; n=len(src)
    state=None; quote=None
    while i<n:
        c=src[i]
        if state is None:
            if c=='/' and i+1<n and src[i+1]=='/':
                while i<n and src[i]!='\n': i+=1
                continue
            if c=='/' and i+1<n and src[i+1]=='*':
                i+=2
                while i+1<n and not (src[i]=='*' and src[i+1]=='/'): 
                    if src[i]=='\n': out.append('\n')
                    i+=1
                i+=2
                continue
            if c in '"\'`':
                quote=c; i+=1
                out.append('""' if c!='`' else '``')
                while i<n:
                    if src[i]=='\\': i+=2; continue
                    if src[i]==quote: i+=1; break
                    if src[i]=='\n': out.append('\n')
                    i+=1
                continue
            out.append(c); i+=1
        else:
            i+=1
    return ''.join(out)

CALL=re.compile(r'console\s*\.\s*(log|error|warn|debug|info|trace)\s*\(')
def sites(path):
    src=open(path,encoding='utf8',errors='replace').read()
    s=strip(src)
    res=[]
    for m in CALL.finditer(s):
        j=m.end(); depth=1
        while j<len(s) and depth>0:
            if s[j]=='(': depth+=1
            elif s[j]==')': depth-=1
            j+=1
        args=s[m.end():j-1]
        # dynamic = anything left after removing "" literals, `` , whitespace, commas, +
        r=args.replace('""','').replace('``','')
        r=re.sub(r'[\s,+]','',r)
        res.append((s[:m.start()].count('\n')+1, m.group(1), bool(r), args.strip()[:120]))
    return res

root=sys.argv[1]
tot=0; dyn=0; per={}
for dirpath,dirnames,files in os.walk(root):
    dirnames[:] = [d for d in dirnames if d not in ('node_modules','test','tests')]
    for f in files:
        if not f.endswith('.js'): continue
        p=os.path.join(dirpath,f)
        rel=os.path.relpath(p,root)
        if not (rel.startswith('lib'+os.sep) or rel=='index.js'): continue
        ss=sites(p)
        if ss:
            per[rel]=(len(ss),sum(1 for x in ss if x[2]))
        tot+=len(ss); dyn+=sum(1 for x in ss if x[2])
print(json.dumps({'total':tot,'dynamic':dyn,'per':per},indent=1,sort_keys=True))
